File: red/src/tools.ts
import * as ansible from "red/ansible";
import { stageDir } from "red/cli";
import { PRESERVE_JINJA_DELIMITERS, contentSpec, type Spec, type Template } from "red/scaffold";
import * as tofu from "red/tofu";
import { runtime } from "red/runtime";
import type { Opts } from "red/workflow";
import { failed } from "red/workflow";
import { registrableDomain } from "package-once-red";
import * as sshConfig from "./ssh-config.ts";
import * as validate from "./validate.ts";

import ansibleLocalCfg from "../resources/tools/ansible-local/ansible.cfg" with { type: "text" };
import ansibleLocalInventory from "../resources/tools/ansible-local/inventory.ini" with { type: "text" };
import ansibleLocalMain from "../resources/tools/ansible-local/main.yml" with { type: "text" };
import ansibleCfg from "../resources/tools/ansible/ansible.cfg" with { type: "text" };
import ansibleMain from "../resources/tools/ansible/main.yml" with { type: "text" };
import ansibleCleanup from "../resources/tools/ansible/cleanup.yml" with { type: "text" };
import ansibleCompose from "../resources/tools/ansible/compose.yml" with { type: "text" };
import ansibleCaddyfile from "../resources/tools/ansible/Caddyfile" with { type: "text" };
import ansibleIngester from "../resources/tools/ansible/ingester.yaml" with { type: "text" };
import ansibleOpamp from "../resources/tools/ansible/opamp.yaml" with { type: "text" };
import ansibleKeeper from "../resources/tools/ansible/keeper.yaml" with { type: "text" };
import ansibleClickhouse from "../resources/tools/ansible/clickhouse.yaml" with { type: "text" };
import ansibleFunctions from "../resources/tools/ansible/functions.yaml" with { type: "text" };
import ansibleSmoke from "../resources/tools/ansible/smoke.sh" with { type: "text" };
import ansibleBackupSh from "../resources/tools/ansible/backup.sh" with { type: "text" };
import ansibleBackupService from "../resources/tools/ansible/backup.service" with { type: "text" };
import ansibleBackupTimer from "../resources/tools/ansible/backup.timer" with { type: "text" };
import dnsMainTf from "../resources/tools/dns/main.tf" with { type: "text" };
import infrastructureDigitaloceanTf from "../resources/tools/infrastructure/digitalocean/main.tf" with { type: "text" };
import infrastructureVultrTf from "../resources/tools/infrastructure/vultr/main.tf" with { type: "text" };

export const infrastructureTool = "signoz-infrastructure";
export const dnsTool = "signoz-dns";
export const ansibleTool = "signoz-ansible";
export const ansibleLocalTool = "signoz-ansible-local";
export const templateOpts = PRESERVE_JINJA_DELIMITERS;

export function toolDir(opts: Opts, tool: string): string {
  return stageDir(opts, tool, { defaultProfile: "signoz" });
}

// The template tree this colour carries, keyed the way green names its
// classpath resources: "<path>/<file>" with dots as directories.
const templates: Record<string, string> = {
  "ansible-local/ansible.cfg": ansibleLocalCfg,
  "ansible-local/inventory.ini": ansibleLocalInventory,
  "ansible-local/main.yml": ansibleLocalMain,
  "ansible/ansible.cfg": ansibleCfg,
  "ansible/main.yml": ansibleMain,
  "ansible/cleanup.yml": ansibleCleanup,
  "ansible/compose.yml": ansibleCompose,
  "ansible/Caddyfile": ansibleCaddyfile,
  "ansible/ingester.yaml": ansibleIngester,
  "ansible/opamp.yaml": ansibleOpamp,
  "ansible/keeper.yaml": ansibleKeeper,
  "ansible/clickhouse.yaml": ansibleClickhouse,
  "ansible/functions.yaml": ansibleFunctions,
  "ansible/smoke.sh": ansibleSmoke,
  "ansible/backup.sh": ansibleBackupSh,
  "ansible/backup.service": ansibleBackupService,
  "ansible/backup.timer": ansibleBackupTimer,
  "dns/main.tf": dnsMainTf,
  "infrastructure/digitalocean/main.tf": infrastructureDigitaloceanTf,
  "infrastructure/vultr/main.tf": infrastructureVultrTf,
};

export function template(path: string, file: string): Template {
  const name = `${path.replaceAll(".", "/")}/${file}`;
  const content = templates[name];
  if (content === undefined) throw new Error(`template not found: ${name}`);
  return { name, content };
}

function spec(source: Template, target: string, data: Opts): Spec {
  return { template: source, target, data, opts: templateOpts };
}

const rawSpec = (target: string, content: string): Spec => contentSpec(target, content);

// The source lists as validate parses them, so the template and the
// validator can never disagree about what an entry is.
export const cidrs = validate.cidrs;

export function credentialEnv(opts: Opts, ...slots: string[]): Record<string, string> | undefined {
  const mapping: Record<string, string> = Object.assign(
    {},
    ...[...slots, "provider-backend"].map((slot) => validate.tofuEnv(opts, slot)),
  );
  const env: Record<string, string> = {};
  for (const [key, envVar] of Object.entries(mapping)) {
    const value = String(opts[key] ?? "");
    if (value.length > 0) env[envVar] = value;
  }
  return Object.keys(env).length > 0 ? env : undefined;
}

export const backendCredentialEnv = (opts: Opts) => credentialEnv(opts);

// What `build` and `--dry-run` render in place of a compute output: the
// documentation address, shaped like the selected provider's real `params` so
// every later stage sees the same keys either way.
export function fallbackParams(opts: Opts): Record<string, unknown> {
  return { provider: opts["provider-compute"], ip: "192.0.2.10", user: "root", sudoer: "root",
    name: validate.computeName(opts) };
}

export function outputParams(result: Opts): Record<string, unknown> | undefined {
  const params = (result["tofu/outputs"] as Record<string, unknown> | undefined)?.params;
  return params && typeof params === "object" ? params as Record<string, unknown> : undefined;
}

// Refuse to hand 192.0.2.10 to Ansible. That is the documentation address the
// credential-free build and dry-run paths render with; on a real converge a
// missing compute output must fail loudly rather than quietly point the whole
// playbook at TEST-NET.
export function resolvedCompute(
  result: Opts,
  fallback: Record<string, unknown>,
  outputs: Record<string, unknown> | undefined,
): Opts {
  if (outputs?.ip) return { ...result, ...fallback, ...outputs };
  return { ...result, "red/exit": 1,
    "red/err": "compute produced no ip output; refusing to converge against the documentation address" };
}

// ---------------------------------------------------------------- compute

// Template values for the compute stage. The name and the source lists are
// resolved here once, so a template interpolates values and never branches on
// which provider it belongs to.
export function infrastructureData(opts: Opts): Opts {
  return {
    ...opts,
    "ssh-keygen": validate.keygen(opts),
    "compute-name": validate.computeName(opts),
    "ssh-sources-hcl": tofu.hclList(cidrs(opts, validate.computeKey(opts, "ssh-sources"))),
    "http-sources-hcl": tofu.hclList(cidrs(opts, validate.computeKey(opts, "http-sources"))),
  };
}

// Providers are selected by template directory, `infrastructure/<provider>/`,
// not by conditionals inside one file; the rendered target is the same
// `main.tf` whichever directory it came from.
export function infrastructureTemplate(opts: Opts): Template {
  return template(`infrastructure.${opts["provider-compute"]}`, "main.tf");
}

export async function infrastructureStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, infrastructureTool);
  const specs = [spec(infrastructureTemplate(opts), `${dir}/main.tf`, infrastructureData(opts))];
  const result = await tofu.tofuWithSpec(opts, specs,
    { dir, env: credentialEnv(opts, "provider-compute") });
  if (failed(result)) return result;
  if (opts["red/event"] === "build") return { ...result, ...fallbackParams(opts) };
  if (opts["red/event"] === "delete") return result;
  return resolvedCompute(result, fallbackParams(opts), outputParams(result));
}

// -------------------------------------------------------------------- dns

// The Cloudflare zone the UI host belongs to (its registrable domain).
export function zone(opts: Opts): string | undefined {
  return registrableDomain(opts["signoz-host"]);
}

export function dnsJson(opts: Opts): string {
  return tofu.constructsJson([
    tofu.construct("resource", "cloudflare_dns_record", "signoz", {
      zone_id: "${data.cloudflare_zone.zone.id}",
      name: opts["signoz-host"], content: opts.ip, type: "A",
      proxied: true, ttl: 1,
    }),
  ]);
}

export async function dnsStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, dnsTool);
  const data: Opts = {
    ...opts,
    ip: opts.ip ?? fallbackParams(opts).ip,
    "signoz-zone": zone(opts),
  };
  const specs = [
    spec(template("dns", "main.tf"), `${dir}/main.tf`, data),
    rawSpec(`${dir}/record.tf.json`, dnsJson(data)),
  ];
  return tofu.tofuWithSpec(opts, specs, { dir, env: credentialEnv(opts, "provider-dns") });
}

// ---------------------------------------------------------- ansible (local)

// Only what a `build` genuinely knows. The address, the user and the alias are
// run-time facts and reach the play as extra-vars instead, so the rendered
// playbook carries no IP and is identical on every workstation (SSH Config
// Standard §6).
export function ansibleLocalData(opts: Opts): Opts {
  return {
    ...opts,
    "ssh-keygen": validate.keygen(opts),
    "ssh-config-identity-file": sshConfig.identityFile(opts),
  };
}

export function ansibleLocalSpecs(opts: Opts): Spec[] {
  const dir = toolDir(opts, ansibleLocalTool);
  const data = ansibleLocalData(opts);
  return [
    spec(template("ansible-local", "ansible.cfg"), `${dir}/ansible.cfg`, data),
    spec(template("ansible-local", "inventory.ini"), `${dir}/inventory.ini`, data),
    spec(template("ansible-local", "main.yml"), `${dir}/main.yml`, data),
  ];
}

// Write or remove the `~/.ssh/config` block. The same playbook serves both
// events; `block_state` is what distinguishes them.
export async function ansibleLocalStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, ansibleLocalTool);
  const isDelete = opts["red/event"] === "delete";
  return ansible.ansibleWithSpec(opts, {
    dir,
    inventory: "inventory.ini",
    playbooks: { create: "main.yml", delete: "main.yml" },
    extraVars: {
      host_alias: sshConfig.hostAlias(opts),
      ip: opts.ip ?? fallbackParams(opts).ip,
      user: opts.user ?? "root",
      block_state: isDelete ? "absent" : "present",
    },
  }, ansibleLocalSpecs(opts));
}

// ---------------------------------------------------------------- ansible

function pretty(value: unknown, indent = 0): string {
  if (Array.isArray(value)) {
    if (value.length === 0) return "[ ]";
    return `[ ${value.map((item) => pretty(item, indent)).join(", ")} ]`;
  }
  if (value !== null && typeof value === "object") {
    const entries = Object.entries(value);
    if (entries.length === 0) return "{ }";
    const pad = " ".repeat(indent + 2);
    return `{\n${entries
      .map(([key, nested]) => `${pad}${JSON.stringify(key)} : ${pretty(nested, indent + 2)}`)
      .join(",\n")}\n${" ".repeat(indent)}}`;
  }
  return JSON.stringify(value ?? null);
}

export function inventory(opts: Opts): string {
  return pretty({
    all: {
      children: {
        signoz: {
          hosts: {
            [String(opts.profile)]: {
              ansible_host: opts.ip ?? "192.0.2.10",
              ansible_user: "root",
            },
          },
        },
      },
    },
  });
}

// Template values for the Ansible stage.
//
// Deliberately carries none of the three operator secrets. They reach the host
// as Ansible `lookup('env', ...)` expressions written literally into main.yml,
// where `preserve-jinja-delimiters` passes them through untouched — routing
// them through this map instead would let the template engine HTML-escape the
// quotes and hand Ansible `&#39;`. The secret therefore exists only in the
// process that needs it: not in `.colors/`, not in a golden, not in this map.
export function ansibleData(opts: Opts): Opts {
  return {
    ...opts,
    ip: opts.ip ?? "192.0.2.10",
    "ssh-keygen": validate.keygen(opts),
  };
}

export function ansibleSpecs(opts: Opts): Spec[] {
  const dir = toolDir(opts, ansibleTool);
  const data = ansibleData(opts);
  const files = ["ansible.cfg", "main.yml", "cleanup.yml", "compose.yml", "Caddyfile",
                 "ingester.yaml", "opamp.yaml", "keeper.yaml", "clickhouse.yaml",
                 "functions.yaml", "smoke.sh", "backup.sh", "backup.service", "backup.timer"];
  return [
    ...files.map((name) => spec(template("ansible", name), `${dir}/${name}`, data)),
    rawSpec(`${dir}/inventory.json`, inventory(data)),
  ];
}

export async function ansibleStep(opts: Opts): Promise<Opts> {
  const dir = toolDir(opts, ansibleTool);
  if (opts["red/event"] === "delete" && !opts.ip) {
    // No compute in state: there is no host to stop, and the cleanup play
    // would only fail against the placeholder address.
    return { ...opts, "red/exit": 0 };
  }
  return ansible.ansibleWithSpec(opts, {
    dir,
    inventory: "inventory.json",
    playbooks: { create: "main.yml", delete: "cleanup.yml" },
    hostKeyChecking: false,
  }, ansibleSpecs(opts));
}

// ------------------------------------------------------------- acceptance

async function run(args: string[]) {
  return runtime.exec(args, { timeoutMs: 20000 });
}

// True once `args` exits zero, retrying every five seconds.
export async function waitFor(args: string[], attempts: number): Promise<boolean> {
  for (let remaining = attempts; ; remaining -= 1) {
    const result = await run(args);
    if (result.exit === 0) return true;
    if (remaining <= 0) return false;
    await Bun.sleep(5000);
  }
}

// The status code a request returns, as a string, or "000" when the request
// never completed.
export async function httpStatus(args: string[]): Promise<string> {
  return String((await run(args)).out ?? "").trim();
}

// Public health checks after a real create.
//
// The end-to-end ingest proof runs on the server, inside the playbook, where
// the generated ingestion token lives. What is checked from here is what the
// internet can reach: the UI over HTTPS, and an OTLP endpoint that refuses an
// unauthenticated write. The refusal is the point — SigNoz community edition
// has no ingestion keys of its own, so an endpoint that accepted this request
// would be an open write path into ClickHouse.
export async function acceptanceStep(opts: Opts): Promise<Opts> {
  if (opts["red/event"] !== "create") return { ...opts, "red/exit": 0 };
  const base = `https://${opts["signoz-host"]}`;
  if (!(await waitFor(["curl", "-fsS", "-o", "/dev/null", `${base}/`], 60))) {
    return { ...opts, "red/exit": 1,
      "red/err": "the SigNoz UI did not become reachable over HTTPS" };
  }
  const health = await httpStatus([
    "curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
    `${base}/api/v1/health`,
  ]);
  const otlp = await httpStatus([
    "curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
    "-X", "POST", "-H", "content-type: application/json",
    "--data", '{"resourceLogs":[]}', `${base}/v1/logs`,
  ]);
  if (health !== "200") {
    return { ...opts, "red/exit": 1,
      "red/err": `the SigNoz API is not healthy: /api/v1/health returned ${health}` };
  }
  if (otlp !== "401") {
    return { ...opts, "red/exit": 1,
      "red/err": "the public OTLP endpoint is not gated: an unauthenticated " +
        `/v1/logs returned ${otlp} rather than 401` };
  }
  return { ...opts, "red/exit": 0,
    "signoz/acceptance": { ui: "ok", health, "otlp-unauthenticated": otlp } };
}

File: blue/src/package_signoz_blue/tools.py
"""The steps and every template spec, the port of io.github.getcolors.signoz.tools."""

from __future__ import annotations

import asyncio
import json
from pathlib import Path
import re

from blue import tofu
from blue.ansible import ansible_with_spec
from blue.cli import stage_dir
from blue.runtime import runtime
from blue.scaffold import PRESERVE_JINJA_DELIMITERS, content_spec
from package_once_blue.utils import registrable_domain

from . import ssh_config, validate

infrastructure_tool = "signoz-infrastructure"
dns_tool = "signoz-dns"
ansible_tool = "signoz-ansible"
ansible_local_tool = "signoz-ansible-local"
ROOT = Path(__file__).parent / "resources"
template_opts = PRESERVE_JINJA_DELIMITERS


def tool_dir(opts: dict, tool: str) -> str:
    return stage_dir(opts, tool, default_profile="signoz")


def template(path: str, file: str) -> dict:
    name = f"tools/{path}/{file}"
    return {"name": name, "content": (ROOT / name).read_text()}


def spec(source: dict, target: str, data: dict) -> dict:
    return {"template": source, "target": target, "data": data, "opts": template_opts}


def raw_spec(target: str, content: str) -> dict:
    return content_spec(target, content)


def cidrs(opts: dict, key: str) -> list[str]:
    value = opts.get(key)
    xs = value if isinstance(value, list) else re.split(
        r"[,\s]+", "" if value is None else str(value))
    return [s for s in (str(x).strip() for x in xs) if s]


def credential_env(opts: dict, *slots: str) -> dict[str, str] | None:
    merged: dict[str, str] = {}
    for slot in [*slots, "provider-backend"]:
        merged.update(validate.tofu_env(opts, slot))
    result = {}
    for key, env_var in merged.items():
        value = "" if opts.get(key) is None else str(opts.get(key))
        if value:
            result[env_var] = value
    return result or None


def backend_credential_env(opts: dict) -> dict[str, str] | None:
    return credential_env(opts)


def fallback_params(opts: dict) -> dict:
    return {"ip": "192.0.2.10", "user": "root", "sudoer": "root",
            "name": opts.get("profile")}


def output_params(result: dict) -> dict | None:
    return (result.get("tofu/outputs") or {}).get("params")


# ---------------------------------------------------------------- compute


def infrastructure_data(opts: dict) -> dict:
    return {**opts,
            "ssh-keygen": validate.keygen(opts),
            "ssh-sources-hcl": tofu.hcl_list(cidrs(opts, "vultr-ssh-sources")),
            "http-sources-hcl": tofu.hcl_list(cidrs(opts, "vultr-http-sources"))}


async def infrastructure_step(opts: dict) -> dict:
    dir = tool_dir(opts, infrastructure_tool)
    specs = [spec(template("infrastructure", "main.tf"), f"{dir}/main.tf",
                  infrastructure_data(opts))]
    result = await tofu.tofu_with_spec(
        opts, specs, dir=dir, env=credential_env(opts, "provider-compute"))
    if (result.get("blue/exit") or 0) > 0:
        return result
    if opts.get("blue/event") == "build":
        return {**result, **fallback_params(opts)}
    if opts.get("blue/event") == "delete":
        return result
    return {**result, **fallback_params(opts), **(output_params(result) or {})}


# -------------------------------------------------------------------- dns


def zone(opts: dict) -> str | None:
    """The Cloudflare zone the UI host belongs to (its registrable domain)."""
    return registrable_domain(opts.get("signoz-host"))


def dns_json(opts: dict) -> str:
    return tofu.constructs_json([
        tofu.construct("resource", "cloudflare_dns_record", "signoz",
                       {"zone_id": "${data.cloudflare_zone.zone.id}",
                        "name": opts.get("signoz-host"),
                        "content": opts.get("ip"), "type": "A",
                        "proxied": True, "ttl": 1})])


async def dns_step(opts: dict) -> dict:
    dir = tool_dir(opts, dns_tool)
    data = {**opts,
            "ip": opts.get("ip") or fallback_params(opts)["ip"],
            "signoz-zone": zone(opts)}
    specs = [spec(template("dns", "main.tf"), f"{dir}/main.tf", data),
             raw_spec(f"{dir}/record.tf.json", dns_json(data))]
    return await tofu.tofu_with_spec(
        opts, specs, dir=dir, env=credential_env(opts, "provider-dns"))


# ---------------------------------------------------------- ansible (local)


def ansible_local_data(opts: dict) -> dict:
    """Only what a `build` genuinely knows. The address, the user and the alias
    are run-time facts and reach the play as extra-vars instead, so the
    rendered playbook carries no IP and is identical on every workstation (SSH
    Config Standard §6)."""
    return {**opts,
            "ssh-keygen": validate.keygen(opts),
            "ssh-config-identity-file": ssh_config.identity_file(opts)}


def ansible_local_specs(opts: dict) -> list[dict]:
    dir = tool_dir(opts, ansible_local_tool)
    data = ansible_local_data(opts)
    return [spec(template("ansible-local", name), f"{dir}/{name}", data)
            for name in ["ansible.cfg", "inventory.ini", "main.yml"]]


async def ansible_local_step(opts: dict) -> dict:
    """Write or remove the `~/.ssh/config` block. The same playbook serves both
    events; `block_state` is what distinguishes them."""
    dir = tool_dir(opts, ansible_local_tool)
    delete = opts.get("blue/event") == "delete"
    return await ansible_with_spec(
        opts, ansible_local_specs(opts),
        dir=dir, inventory="inventory.ini",
        playbooks={"create": "main.yml", "delete": "main.yml"},
        extra_vars={"host_alias": ssh_config.host_alias(opts),
                    "ip": opts.get("ip") or fallback_params(opts)["ip"],
                    "user": opts.get("user") or "root",
                    "block_state": "absent" if delete else "present"})


# ---------------------------------------------------------------- ansible


def _pretty(value, indent=0):
    """Cheshire's pretty JSON, byte for byte — Green's artifact contract."""
    if isinstance(value, list):
        if not value:
            return "[ ]"
        return "[ " + ", ".join(_pretty(item, indent) for item in value) + " ]"
    if isinstance(value, dict):
        if not value:
            return "{ }"
        pad = " " * (indent + 2)
        body = ",\n".join(f"{pad}{json.dumps(str(k))} : {_pretty(v, indent + 2)}"
                          for k, v in value.items())
        return "{\n" + body + "\n" + " " * indent + "}"
    return json.dumps(value)


def inventory(opts: dict) -> str:
    return _pretty(
        {"all": {"children": {"signoz": {"hosts": {
            opts.get("profile"): {"ansible_host": opts.get("ip") or "192.0.2.10",
                                  "ansible_user": "root"}}}}}})


def ansible_data(opts: dict) -> dict:
    """Template values for the Ansible stage.

    Deliberately carries none of the three operator secrets. They reach the
    host as Ansible `lookup('env', ...)` expressions written literally into
    main.yml, where `preserve-jinja-delimiters` passes them through untouched —
    routing them through this map instead would let the template engine
    HTML-escape the quotes and hand Ansible `&#39;`. The secret therefore
    exists only in the process that needs it: not in `.colors/`, not in a
    golden, not in this map."""
    return {**opts,
            "ip": opts.get("ip") or "192.0.2.10",
            "ssh-keygen": validate.keygen(opts)}


ANSIBLE_FILES = [
    "ansible.cfg", "main.yml", "cleanup.yml", "compose.yml", "Caddyfile",
    "ingester.yaml", "opamp.yaml", "keeper.yaml", "clickhouse.yaml",
    "functions.yaml", "smoke.sh", "backup.sh", "backup.service", "backup.timer",
]


def ansible_specs(opts: dict) -> list[dict]:
    dir = tool_dir(opts, ansible_tool)
    data = ansible_data(opts)
    return [*[spec(template("ansible", name), f"{dir}/{name}", data)
              for name in ANSIBLE_FILES],
            raw_spec(f"{dir}/inventory.json", inventory(data))]


async def ansible_step(opts: dict) -> dict:
    dir = tool_dir(opts, ansible_tool)
    if opts.get("blue/event") == "delete" and not opts.get("ip"):
        # No compute in state: there is no host to stop, and the cleanup play
        # would only fail against the placeholder address.
        return {**opts, "blue/exit": 0}
    return await ansible_with_spec(
        opts, ansible_specs(opts),
        dir=dir, inventory="inventory.json",
        playbooks={"create": "main.yml", "delete": "cleanup.yml"},
        host_key_checking=False)


# ------------------------------------------------------------- acceptance


async def wait_for(args: list[str], attempts: int) -> bool:
    """True once `args` exits zero, retrying every five seconds."""
    n = attempts
    while True:
        result = await runtime.exec(args, timeout_ms=20000)
        if result.exit == 0:
            return True
        if n > 0:
            await asyncio.sleep(5)
            n -= 1
        else:
            return False


async def run(args: list[str]):
    return await runtime.exec(args, timeout_ms=20000)


async def http_status(args: list[str]) -> str:
    """The status code a request returns, as a string, or "000" when the
    request never completed."""
    return str((await run(args)).out or "").strip()


async def acceptance_step(opts: dict) -> dict:
    """Public health checks after a real create.

    The end-to-end ingest proof runs on the server, inside the playbook, where
    the generated ingestion token lives. What is checked from here is what the
    internet can reach: the UI over HTTPS, and an OTLP endpoint that refuses an
    unauthenticated write. The refusal is the point — SigNoz community edition
    has no ingestion keys of its own, so an endpoint that accepted this request
    would be an open write path into ClickHouse."""
    if opts.get("blue/event") != "create":
        return {**opts, "blue/exit": 0}
    base = f"https://{opts.get('signoz-host')}"
    if not await wait_for(["curl", "-fsS", "-o", "/dev/null", f"{base}/"], 60):
        return {**opts, "blue/exit": 1,
                "blue/err": "the SigNoz UI did not become reachable over HTTPS"}
    health = await http_status(
        ["curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
         f"{base}/api/v1/health"])
    otlp = await http_status(
        ["curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
         "-X", "POST", "-H", "content-type: application/json",
         "--data", '{"resourceLogs":[]}', f"{base}/v1/logs"])
    if health != "200":
        return {**opts, "blue/exit": 1,
                "blue/err": f"the SigNoz API is not healthy: /api/v1/health returned {health}"}
    if otlp != "401":
        return {**opts, "blue/exit": 1,
                "blue/err": ("the public OTLP endpoint is not gated: an unauthenticated "
                             f"/v1/logs returned {otlp} rather than 401")}
    return {**opts, "blue/exit": 0,
            "signoz/acceptance": {"ui": "ok", "health": health,
                                  "otlp-unauthenticated": otlp}}
